### PR TITLE
Add a warning if tmpdir/rundir are not on tmpfs

### PR DESCRIPTION
**Pull request by mheon**
Podman requires several directories to be on tmpfs so we can properly detect (and clean up after) reboots: the tmpdir (for Libpod files) and runroot (for c/storage files). We have not done any validation that they are actual on tmpfs filesystems in the past, which leads to no end of people wondering why Podman is broken on their system (where, perhaps, /tmp is not a tmpfs, which the FHS allows because sanity cannot exist in this world, and there is no systemd session so /run/user/$UID does not exist).

This adds an actual warning on launching Podman with such an unsupported configuration. There's an environment variable if the user is certain they want to run Podman like this (perhaps having used systemd-tmpfiles to automatically clean the relevant dir on reboot), but otherwise we warn every time a command is run with runroot/tmpdir on a non-tmpfs filesystem.

#### Does this PR introduce a user-facing change?

```release-note
Podman now warns if a temporary files directory is not placed on a tmpfs filesystem.
```

### CODE REVIEW

File: libpod/runtime_linux.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"os"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+// Print a warning if a path is not on a tmpfs.
+func warnIfNotTmpfs(paths []string) {
+	envVar := "PODMAN_SUPPRESS_TMPFS_WARNING"
```
Comment (Luap99) [inline libpod/runtime_linux.go:14]: define this as const

**Comment by afbjorklund** [comment]
> perhaps, /tmp is not a tmpfs, which the FHS allows because sanity cannot exist in this world, and there is no systemd session so /run/user/$UID does not exist)

I thought Podman still supported installations with a traditional `/tmp` filesystem and a traditional `init` system...

File: libpod/runtime_linux.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+	notTmpfs := make([]string, 0, len(paths))
+
+	for _, path := range paths {
+		statfs := new(unix.Statfs_t)
+		if err := unix.Statfs(path, statfs); err != nil {
+			return
+		}
+
+		if statfs.Type != unix.TMPFS_MAGIC {
```
Comment (Luap99) [inline libpod/runtime_linux.go:27]: I think there also the less common ramfs so I think we should check that here to.

**Review by Luap99 — COMMENTED** [review]
small comments, all I can say is we should have done this much sooner given the amount of problems we had with this

**Comment by Luap99** [comment]
> > perhaps, /tmp is not a tmpfs, which the FHS allows because sanity cannot exist in this world, and there is no systemd session so /run/user/$UID does not exist)
> 
> I thought Podman still supported installations with a traditional `/tmp` filesystem and a traditional `init` system...

We do, but that doesn't change the fact that we depend on tmpfs or some other util to remove these paths on boot.

**Comment by mheon** [comment]
@edsantiago The system tests don't use put the test directory on tmpfs, right? If so we can test this on tmpfs

**Comment by baude** [comment]
LGTM, do any docs need updates or should have clarification?

**Comment by edsantiago** [comment]
> @edsantiago The system tests don't use put the test directory on tmpfs, right? If so we can test this on tmpfs

@mheon I've been struggling to understand this. The best answer I can give is, (for the most part) system tests just invoke `podman` with no special magic args. There are a few very specific cases in which there are `--root --runroot --tmpdir`. Does that address your question?

File: libpod/runtime_linux.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+const TMPFS_WARNING_ENV_VAR = "PODMAN_SUPPRESS_TMPFS_WARNING"
+
+// Print a warning if a path is not on a tmpfs.
+func warnIfNotTmpfs(paths []string) {
+	if os.Getenv(TMPFS_WARNING_ENV_VAR) != "" {
```
Comment (giuseppe) [inline libpod/runtime_linux.go:16]: I think we need to avoid the warning also when running in a container (the `container` env is set)

**Comment by giuseppe** [comment]
could we instead write a file to the tmpdir/rundir and with the value of `/proc/sys/kernel/random/boot_id` and compare it each time?  If it has a different value, then we return an error instead of a warning.

**Comment by mheon** [comment]
@giuseppe That'd work for generic reboot protection, but it wouldn't help clear files afterwards - this PR was motivated by the IPAM DB not being removed on reboot leading to us running out of IP addresses to allocate. We could theoretically build that entirely into Podman as well, have a list of files which must be removed if reboot is detected, but that seems overly complex when we can just make sure a tmpfs is in use?

**Comment by mheon** [comment]
@edsantiago Yeah that should be fine, thanks. Needed to make sure that we're using standard paths and not throwing everything on a tmpfs for performance or similar.

**Comment by giuseppe** [comment]
> @giuseppe That'd work for generic reboot protection, but it wouldn't help clear files afterwards - this PR was motivated by the IPAM DB not being removed on reboot leading to us running out of IP addresses to allocate. We could theoretically build that entirely into Podman as well, have a list of files which must be removed if reboot is detected, but that seems overly complex when we can just make sure a tmpfs is in use?

I was thinking about raising an error and terminating the Podman process if the content of `/proc/sys/kernel/random/boot_id` is different than `$RUNDIR/boot_id` with something like `the content of $FILE1 is different than $FILE2, please remove $RUNDIR and try again`

**Comment by mheon** [comment]
Hm. I like the idea, but I'd also like to backport this patch, and I don't know if introducing a new error in existing releases is a good idea. Maybe I should target this at 5.0 only, and we make it an error in main?

**Comment by giuseppe** [comment]
> Hm. I like the idea, but I'd also like to backport this patch, and I don't know if introducing a new error in existing releases is a good idea. Maybe I should target this at 5.0 only, and we make it an error in main?

sure, that works for me

**Comment by mheon** [comment]
Update: None of our VMs actually use tmpfs /tmp so this breaks absolutely everything.

**Comment by mheon** [comment]
Passed the ignore-warning environment variable through to all Cirrus tasks. Good news is that it makes the test easier; just unset that environment variable and run Podman.

**Comment by edsantiago** [comment]
https://github.com/containers/automation_images/pull/340 will make `/tmp` be tmpfs, but that won't help with this v5.0 PR.

**Comment by mheon** [comment]
The PR against main will probably be completely different, and not need the environment variable at all.
Fortunately v5.0 is not going to be an LTS branch so we can stop caring about this little idiosyncrasy in a few months when 5.1 comes out.

File: test/system/005-info.bats [hunk 4]
```diff
@@ -258,4 +258,18 @@ EOF
...
+    skip_if_remote "Check requires a local Libpod"
+
+    # We need a directory that is not tmpfs to put Podman in
+    OLDTMPFS=$PODMAN_SUPPRESS_TMPFS_WARNING
+    unset PODMAN_SUPPRESS_TMPFS_WARNING
+
+    # Anything that creates a Libpod should trigger the warning.
+    run_podman info
+    assert "$output" =~ "The following temporary files directories are not on a tmpfs filesystem"
+
```
Comment (edsantiago) [inline test/system/005-info.bats:271]: Might I also suggest
```bash
    tmpfs_tmp=$(mktemp -d /dev/shm)
    run_podman --tmpdir=$tmpfs_tmp/tmp --runroot=$tmpfs_tmp/runroot --root=$tmpfs_tmp/root info
    assert "$output" !~ "are not on a tmpfs"
```

Comment (mheon) [reply test/system/005-info.bats:271]: Any reason to not just `mktemp` and then mount a tmpfs over that directory?

Comment (edsantiago) [reply test/system/005-info.bats:271]: Laziness? And, will that work rootless? I've never seen a system without `/dev/shm` (at least not in the last 20 years)

Comment (mheon) [reply test/system/005-info.bats:271]: Fair point on rootless

**Comment by mheon** [comment]
@edsantiago Does the test logic seem OK to you? It's acting like the environment variable is still defined.

**Comment by edsantiago** [comment]
Maybe because `/run`?
```
runRoot: /run/containers/storage
```
I don't see tmpdir, though. Maybe that's in XDG?

I still think it's better to test with explicit `--runroot/--tmpdir` + `/dev/shm / /var/tmp`

**Comment by Luap99** [comment]
Why is this targeted against the 5.0 branch? I don't think it is a good idea to backport such a big warning in a patch release.

**Comment by mheon** [comment]
We decided on an alternate implementation for Main - we'll be outright refusing to run if we detect an unhandled reboot in 5.1. This was lower-impact so I thought it would still be useful in 5.0.1. I'm OK with leaving it out until we discuss further when you get back, though.

**Comment by Luap99** [comment]
Note I feel rather strongly about not adding such big warnings in a patch release, we lived long enough with the status quo so waiting few more weeks shouldn't be that big of a deal. And if we plan to fix this in a different way on main anyway then I think we should just do that and not introduce two different fixes in two different versions which will likely cause some differences in behavior and thus confusion.

**Comment by mheon** [comment]
I'll go ahead and close, then. We can fix properly in 5.1.

**Pull request closed without merging**